### PR TITLE
[deliver] Fix builds whose versions have components with leading zeros infinitely wait for submission

### PR DESCRIPTION
**Pull request by justindhill**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
This PR fixes an issue in Deliver's `SubmitForReview` which causes builds containing zero-padding in their version components to fail to submit. It is caused by an Apple issue which causes the App Store Connect API to temporarily return a `train_version` whose components are stripped of their zero padding. For example, "2019.04.01" would be returned as "2019.4.1". During the processing, eventually the API starts returning the correct form of the version again, i.e. "2019.04.01". 

The above causes Deliver to infinitely wait for the build to be processed because it notes the most recently uploaded build as soon as Transporter finishes and uses it to compare the version of builds in future calls to `candidate_builds` to prevent accidentally selecting the wrong build for submission.

Below is some trimmed json demonstrating the relevant changes to the data throughout processing:

Immediately after upload:
```json
{
	"buildVersion": "82",
	"trainVersion": "2019.4.1",
	"uploadDate": 1554160879000,
	"processing": true,
	"processingState": "processing",
	"processingError": null,
}
```

After the first phase of processing; note the trainVersion format has changed:
```json
{
	"buildVersion": "82",
	"trainVersion": "2019.04.01",
	"uploadDate": 1554161105000,
	"processing": true,
	"processingState": "processing",
	"processingError": null,
}
```

Processing complete:
```json
{
	"buildVersion": "82",
	"trainVersion": "2019.04.01",
	"uploadDate": 1554161105000,
	"processing": false,
	"processingState": null,
	"processingError": null,
}
```

Should solve fastlane/fastlane#13372 and fastlane/fastlane#13577

### Solution

The solution is to use the `app_version` attached to the options hash for comparing to the `candidate_builds` instead of the `train_version` from the previously found build. This ensures that a correctly formatted version is always used for comparison. In cases where an `app_version` is not supplied in the options hash, such as when calling `SubmitForReview` without an ipa and also not providing an `app_version` explicitly, we fall back to the behavior of using the one on the first build we find and hoping for the best.

I added a test that confirms that `wait_for_build` properly returns a build with the supplied app version even when `candidate_builds` returns a build with the zeros stripped. I also added a test that confirms a build is returned when nil is passed as the app version to `wait_for_build`.

### CODE REVIEW

**Comment by justindhill** [comment]
I signed it.

**Comment by justindhill** [comment]
Ok, signed with corporate email as well.

**Comment by justindhill** [comment]
Added my corporate email to GitHub.

File: deliver/spec/submit_for_review_spec.rb
Comment (janpio) [inline deliver/spec/submit_for_review_spec.rb]: Should this maybe moved to `make_fake_version`?

**Comment by janpio** [comment]
Oh wow, bad Apple - and all developers can pretty much guess what is going on there 🙈 

Fix looks good on a superficial review ✔️

File: deliver/spec/submit_for_review_spec.rb
Comment (justindhill) [reply deliver/spec/submit_for_review_spec.rb]: Sure, that makes sense. I'll change it right now.

Comment (justindhill) [reply deliver/spec/submit_for_review_spec.rb]: Done

File: deliver/spec/submit_for_review_spec.rb
Comment (giginet) [inline deliver/spec/submit_for_review_spec.rb]: You can use rspec's `double` instead.

File: deliver/spec/submit_for_review_spec.rb
Comment (giginet) [inline deliver/spec/submit_for_review_spec.rb]: You should use `double` and `allow`.

```ruby
build = double('fake_build')
allow(build).to receive(:[]).with(:train_version).and_return("1.2.3")
allow(build).to receive(:[]).with(:build_version).and_return("1")
allow(build).to receive(:[]).with(:processing).and_return(false)
[build]
```

File: deliver/lib/deliver/submit_for_review.rb
Comment (giginet) [inline deliver/lib/deliver/submit_for_review.rb]: You can use `&.`.

```suggestion
        latest_build_matches = (latest_build&.train_version == app_version && latest_build&.build_version == build&.build_version)
```

File: deliver/spec/submit_for_review_spec.rb
Comment (justindhill) [reply deliver/spec/submit_for_review_spec.rb]: I don't really know Ruby, so I'm totally willing to defer to your expertise, but would you mind explaining the advantage of stubbing with rspec as opposed to just setting the value? To my untrained eye, it just looks more verbose while accomplishing the same thing.

File: deliver/lib/deliver/submit_for_review.rb
Comment (justindhill) [reply deliver/lib/deliver/submit_for_review.rb]: This line just extracts the expression from the line below, but what you proposed is a great syntax I didn't know about!

File: deliver/spec/submit_for_review_spec.rb
Comment (giginet) [reply deliver/spec/submit_for_review_spec.rb]: According to RSpec documentation.
https://relishapp.com/rspec/rspec-mocks/docs/basics/test-doubles

`any message you have not allowed or expected will trigger an error`

I think that is the main benefit using Test Doubles.
Using test doubles, we can notice unexpected method callings.

File: deliver/lib/deliver/submit_for_review.rb
Comment (giginet) [reply deliver/lib/deliver/submit_for_review.rb]: I'm sorry, this syntax is not allowed on fastlane project 😢 
`&.` was introduced since Ruby 2.3, but unfortunately, TargetRubyVersion is still 2.0.
https://github.com/fastlane/fastlane/blob/master/.rubocop.yml#L244

This syntax is correct, but we can't use this.
https://circleci.com/gh/fastlane/fastlane/30932?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Comment (giginet) [reply deliver/lib/deliver/submit_for_review.rb]: `build` seems to be optional, so we have to do nil-cheking for `build`.

`latest_build_matches = (latest_build && build && latest_build.train_version == app_version && latest_build.build_version == build.build_version)`

**Comment by justindhill** [comment]
@giginet @janpio I've made some changes based on your feedback and they're ready for review.

1. I made wait_for_build read a bit more nicely in addition to addressing the nil check @giginet pointed out.
2. I reworked the tests to use doubles instead of OpenStructs.

File: deliver/lib/deliver/submit_for_review.rb
Comment (giginet) [inline deliver/lib/deliver/submit_for_review.rb]: you can add `private`

File: deliver/spec/submit_for_review_spec.rb
Comment (giginet) [inline deliver/spec/submit_for_review_spec.rb]: You can drop `return`.

File: deliver/spec/submit_for_review_spec.rb
Comment (giginet) [inline deliver/spec/submit_for_review_spec.rb]: You can drop `return` in ruby way.

**Review by giginet — APPROVED** [review]
Great job 🌸

File: deliver/lib/deliver/submit_for_review.rb
Comment (justindhill) [reply deliver/lib/deliver/submit_for_review.rb]: Sure, done.

File: deliver/spec/submit_for_review_spec.rb
Comment (justindhill) [reply deliver/spec/submit_for_review_spec.rb]: Done.

File: deliver/spec/submit_for_review_spec.rb
Comment (justindhill) [reply deliver/spec/submit_for_review_spec.rb]: Done.

**Comment by giginet** [comment]
@justindhill Please agree with CLA before merging

https://github.com/fastlane/fastlane/pull/14513#issuecomment-478844478

**Comment by justindhill** [comment]
> @justindhill Please agree with CLA before merging
> 
> [#14513 (comment)](https://github.com/fastlane/fastlane/pull/14513#issuecomment-478844478)

@giginet I actually did agree to the CLA. The PR was labeled with cla:yes before I accepted your suggestion a few days ago, at which point @googlebot removed the label. According to the comment it left, you have to manually re-add the tag after confirming consent from all developers. Since you are the other developer, I think it's ok. :)

**Comment by janpio** [comment]
Yep, @googlebot is not smart enough for Github suggestions that are accepted, but the `cla:no` will be no problem here when someone decides this is good to merge - so nothing to do in that regard.

File: deliver/lib/deliver/submit_for_review.rb
Comment (vincentisambart) [inline deliver/lib/deliver/submit_for_review.rb]: That is incorrect, the `app_version` passed to `deliver` is `build_version` not `train_version`.

**Pull request merged**
